# APRIORI ALGORİTMASI KULLANARAK BİRLİKTELİK KURAL ANALİZİ 

X: 1.ürün Y: 2.ürün N: Toplam Alışveriş

- Destek (Support) : Ürünlerin Birlikte Geçme Olasılığı

Support(X, Y) = Freq(X,Y)/N

( X ve Y farklı ürünlerinin birlikte geçme frekansı / Toplam Alışveriş )

Support değerini hesaplama amacımız eşik değeri belirlemek. Veri setine bakıldığında bir çok ürün beraber görülebilir eşik değerinin altında kalan ürünlerde eleme yapacağız.

- Güven (Confidence) : X’i Alanların Y’yi Alma Olasılığı

Confidence(X, Y) = Freq(X,Y)/Freq(X)

( X ve Y farklı ürünlerinin birlikte geçme frekansı / X’in gözlenme frekansı )

Analiz yapıldığında ister support değerine göre ister confidence değerine göre eşik değeri belirlenebilir. Eşik değerini gerçek hayat verilerinde çok az alabiliriz.

- Lift = Support(X,Y)/(Support(X)*Support(Y))

X ürünü alanların Y ürünü satın alması şu kadar kat artıyor yorumu vardır.

In [21]:
# .csv formatındaki veri setini okuma işlemi
# Süt,Ekmek,Bisküvü olan sutün başlığını names=['Ürünler'] ile ürünlere çevirdik.

import pandas as pd
df = pd.read_csv("ürünler.csv", names=['ürünler'], header = None)
df.head()

ürünler
0                SÜT,EKMEK,BİSKÜVİ
1  EKMEK,SÜT,BİSKÜVİ,MISIR GEVREĞİ
2                 EKMEK,ÇAY,PEYNİR
3           REÇEL,ZEYTİN,EKMEK,SÜT
4               ZEYTİN,ÇAY,BİSKÜVİ

In [22]:
# 20 satır ve 1 sütun var.

df.shape

(20, 1)

In [23]:
# Sutün adı ve tipi.

df.columns

Index(['ürünler'], dtype='object')

In [10]:
# Her satırda gözlenen gözlemler

df.values

array([['SÜT,EKMEK,BİSKÜVİ'],
       ['EKMEK,SÜT,BİSKÜVİ,MISIR GEVREĞİ'],
       ['EKMEK,ÇAY,PEYNİR'],
       ['REÇEL,ZEYTİN,EKMEK,SÜT'],
       ['ZEYTİN,ÇAY,BİSKÜVİ'],
       ['EKMEK,ÇAY,PEYNİR'],
       ['ZEYTİN,ÇAY,MISIR GEVREĞİ'],
       ['ZEYTİN,EKMEK,ÇAY,BİSKÜVİ'],
       ['REÇEL,ZEYTİN,EKMEK,ÇAY'],
       ['EKMEK,SÜT'],
       ['KAHVE,KOLA,BİSKÜVİ,MISIR GEVREĞİ'],
       ['KAHVE,KOLA,BİSKÜVİ,MISIR GEVREĞİ'],
       ['KAHVE,ŞEKER,PEYNİR'],
       ['EKMEK,KAHVE,KOLA'],
       ['EKMEK,ŞEKER,BİSKÜVİ'],
       ['KAHVE,ŞEKER,MISIR GEVREĞİ'],
       ['EKMEK,ŞEKER,PEYNİR'],
       ['EKMEK,KAHVE,ŞEKER'],
       ['EKMEK,KAHVE,ŞEKER'],
       ['ÇAY,SÜT,KAHVE,MISIR GEVREĞİ']], dtype=object)

In [11]:
# Her satırda tek bir gözlem birimi varmış gibiydi bunu virgül ile tek tek ayırma işlemi yaptık.
data = list(df["ürünler"].apply(lambda x:x.split(',')))
data 



[['SÜT', 'EKMEK', 'BİSKÜVİ'],
 ['EKMEK', 'SÜT', 'BİSKÜVİ', 'MISIR GEVREĞİ'],
 ['EKMEK', 'ÇAY', 'PEYNİR'],
 ['REÇEL', 'ZEYTİN', 'EKMEK', 'SÜT'],
 ['ZEYTİN', 'ÇAY', 'BİSKÜVİ'],
 ['EKMEK', 'ÇAY', 'PEYNİR'],
 ['ZEYTİN', 'ÇAY', 'MISIR GEVREĞİ'],
 ['ZEYTİN', 'EKMEK', 'ÇAY', 'BİSKÜVİ'],
 ['REÇEL', 'ZEYTİN', 'EKMEK', 'ÇAY'],
 ['EKMEK', 'SÜT'],
 ['KAHVE', 'KOLA', 'BİSKÜVİ', 'MISIR GEVREĞİ'],
 ['KAHVE', 'KOLA', 'BİSKÜVİ', 'MISIR GEVREĞİ'],
 ['KAHVE', 'ŞEKER', 'PEYNİR'],
 ['EKMEK', 'KAHVE', 'KOLA'],
 ['EKMEK', 'ŞEKER', 'BİSKÜVİ'],
 ['KAHVE', 'ŞEKER', 'MISIR GEVREĞİ'],
 ['EKMEK', 'ŞEKER', 'PEYNİR'],
 ['EKMEK', 'KAHVE', 'ŞEKER'],
 ['EKMEK', 'KAHVE', 'ŞEKER'],
 ['ÇAY', 'SÜT', 'KAHVE', 'MISIR GEVREĞİ']]

# Data Preprocessing

In [ ]:
#!pip install mlxtend

In [12]:
from mlxtend.preprocessing import TransactionEncoder

In [13]:
#Veri setini istenilen True-False array'ine çevirdik.

te = TransactionEncoder()
te_data = te.fit(data).transform(data)
df = pd.DataFrame(te_data,columns=te.columns_)
df

BİSKÜVİ  EKMEK  KAHVE   KOLA  MISIR GEVREĞİ  PEYNİR  REÇEL    SÜT  ZEYTİN  \
0      True   True  False  False          False   False  False   True   False   
1      True   True  False  False           True   False  False   True   False   
2     False   True  False  False          False    True  False  False   False   
3     False   True  False  False          False   False   True   True    True   
4      True  False  False  False          False   False  False  False    True   
5     False   True  False  False          False    True  False  False   False   
6     False  False  False  False           True   False  False  False    True   
7      True   True  False  False          False   False  False  False    True   
8     False   True  False  False          False   False   True  False    True   
9     False   True  False  False          False   False  False   True   False   
10     True  False   True   True           True   False  False  False   False   
11     True  False   True   True           True   False  False  False   False   
12    False  False   True  False          False    True  False  False   False   
13    False   True   True   True          False   False  False  False   False   
14     True   True  False  False          False   False  False  False   False   
15    False  False   True  False           True   False  False  False   False   
16    False   True  False  False          False    True  False  False   False   
17    False   True   True  False          False   False  False  False   False   
18    False   True   True  False          False   False  False  False   False   
19    False  False   True  False           True   False  False   True   False   

      ÇAY  ŞEKER  
0   False  False  
1   False  False  
2    True  False  
3   False  False  
4    True  False  
5    True  False  
6    True  False  
7    True  False  
8    True  False  
9   False  False  
10  False  False  
11  False  False  
12  False   True  
13  False  False  
14  False   True  
15  False   True  
16  False   True  
17  False   True  
18  False   True  
19   True  False

# Data Association Rules

In [14]:
from mlxtend.frequent_patterns import apriori

In [17]:
# Apriori Fonksiyonunu deneyelim.
# df = True-False array ya da One Hot Encoding ile dönüştürülmüş dataframe
# min_support = Tüm kombinasyonların support değerini istemiyorum bu yüzden belirli bir eşik değerini getir.
# use_colnames = True, sutün isimlerini göster.
# verbose = 1, Toplam kombinasyon sayısını verir.

df1 = apriori(df, min_support=0.02, use_colnames=True, verbose = 1)
df1

Processing 35 combinations | Sampling itemset size 54


support                               itemsets
0      0.35                              (BİSKÜVİ)
1      0.65                                (EKMEK)
2      0.40                                (KAHVE)
3      0.15                                 (KOLA)
4      0.30                        (MISIR GEVREĞİ)
..      ...                                    ...
78     0.05          (ZEYTİN, EKMEK, ÇAY, BİSKÜVİ)
79     0.10  (MISIR GEVREĞİ, BİSKÜVİ, KAHVE, KOLA)
80     0.05            (ZEYTİN, EKMEK, SÜT, REÇEL)
81     0.05            (ZEYTİN, EKMEK, ÇAY, REÇEL)
82     0.05       (ÇAY, MISIR GEVREĞİ, KAHVE, SÜT)

[83 rows x 2 columns]

In [ ]:
''' Bisküvi tüm alışverişlerin % 35'inde, 
    Ekmek tüm alışverişlerin %65'inde veya 
    Çay, Zeytin, Ekmek, Bisküvi tüm satışların % 5'inde beraber yorumları yapılır.''' 

In [18]:
#Alışverişlerde en çok alınan ürünleri yorumlayabilmek için veriyi büyükten küçüğe sıralayabiliriz.

df1.sort_values(by="support", ascending=False)

support                          itemsets
1      0.65                           (EKMEK)
2      0.40                           (KAHVE)
0      0.35                         (BİSKÜVİ)
9      0.35                             (ÇAY)
4      0.30                   (MISIR GEVREĞİ)
..      ...                               ...
50     0.05             (EKMEK, ÇAY, BİSKÜVİ)
51     0.05           (EKMEK, BİSKÜVİ, ŞEKER)
55     0.05     (MISIR GEVREĞİ, BİSKÜVİ, SÜT)
57     0.05              (EKMEK, KAHVE, KOLA)
82     0.05  (ÇAY, MISIR GEVREĞİ, KAHVE, SÜT)

[83 rows x 2 columns]

In [16]:
from mlxtend.frequent_patterns import association_rules

In [19]:
# Aprioride "support" hesabı yapabiliriz, "confidence ve diğerleri" için association rules kullanıyoruz.

association_rules(df1, metric = "confidence", min_threshold = 0.60)[['antecedents', 'consequents','support','confidence']]



antecedents             consequents  support  confidence
0                        (KOLA)               (BİSKÜVİ)     0.10    0.666667
1                      (PEYNİR)                 (EKMEK)     0.15    0.750000
2                       (REÇEL)                 (EKMEK)     0.10    1.000000
3                         (SÜT)                 (EKMEK)     0.20    0.800000
4                      (ZEYTİN)                 (EKMEK)     0.15    0.600000
..                          ...                     ...      ...         ...
81            (ÇAY, KAHVE, SÜT)         (MISIR GEVREĞİ)     0.05    1.000000
82  (MISIR GEVREĞİ, KAHVE, SÜT)                   (ÇAY)     0.05    1.000000
83                 (ÇAY, KAHVE)    (MISIR GEVREĞİ, SÜT)     0.05    1.000000
84                   (ÇAY, SÜT)  (MISIR GEVREĞİ, KAHVE)     0.05    1.000000
85                 (KAHVE, SÜT)    (ÇAY, MISIR GEVREĞİ)     0.05    1.000000

[86 rows x 4 columns]

In [19]:
''' antecedent support ; Birincinin tek başına görülme olasılığı,
    consequent support ; İkincinin tek başına görülme olasılığı,
    support ; İkisinin birlikte görülme olasılığı,
    confidence ; İlki satıldığında ikinci ürünün satılma olasılığı,
    lift ;  İlki satıldığında ikinci ürünün satılma olasılığı şu kadar kat arttı yorumu

'''

' antecedent support ; Birincinin tek başına görülme olasılığı,\n    consequent support ; İkincinin tek başına görülme olasılığı,\n    support ; İkisinin birlikte görülme olasılığı,\n    confidence ; İlki satıldığında ikinci ürünün satılma olasılığı,\n    lift ;  İlki satıldığında ikinci ürünün satılma olasılığı şu kadar kat arttı yorumu\n\n'

## Selecting and Filtering Results 

In [ ]:
''' "confidence" göz önünde bulundurulması gereken tek metrik değil. 
Bir kaç metriğin kombinasyonuda alınabilir.
'''

In [20]:
rules = association_rules(df1, metric = "confidence", min_threshold = 0.6)
rules[ (rules['confidence'] >= 0.6) & (rules['support'] >= 0.2) ]

antecedents consequents  antecedent support  consequent support  \
3             (SÜT)     (EKMEK)                0.25                0.65   
5           (ŞEKER)     (EKMEK)                0.30                0.65   
7   (MISIR GEVREĞİ)     (KAHVE)                0.30                0.40   
8           (ŞEKER)     (KAHVE)                0.30                0.40   
11         (ZEYTİN)       (ÇAY)                0.25                0.35   

    support  confidence      lift  leverage  conviction  zhangs_metric  
3       0.2    0.800000  1.230769    0.0375        1.75       0.250000  
5       0.2    0.666667  1.025641    0.0050        1.05       0.035714  
7       0.2    0.666667  1.666667    0.0800        1.80       0.571429  
8       0.2    0.666667  1.666667    0.0800        1.80       0.571429  
11      0.2    0.800000  2.285714    0.1125        3.25       0.750000

# Reporting

İndirgediğim veriseti üzerinden analiz yapıyor olacağım.
- Alışverişlerde Zeytinin'in tek başına görülme olasığı %25, Çay'ın tek başına görülme olasılığı %35.
- 100 alışverişin 20'sinde mutlaka Zeytinin ve Çay beraber satın alınıyor.
- Zeytin satıldığında çay satılma olasılığı 0.800.. yani %80.
- Zeytin satılan satışlarda çay satılma olasılığı 2.28 kat artmaktadır.


Aksiyon Fikri: 
- zeytin alan biri %80 gibi yüksek bir ihtimalle çay almaktadır ve çay satışını 2.28 artırmaktadır. Bu iki ürün birbirinden uzak yerlerde konumlandırılarak müşterinin market içi dolaşması sağlanabilir, bu süreçte müşteri diğer ürünlere göz atabilir ve yahut satın alabilir. 
